distributed/deploy/spec.py

```diff
@@ -0,0 +1,296 @@
+    needs to be called with the provided options, support ``__await__`` and
+    ``close`` methods and the ``worker_address`` property..
+
+    Also note that uniformity of the specification is not required.  Other API
+    could be added externally (in subclasses) that adds workers of different
+    specifications into the same dictionary.
+    """
+
+    def __init__(
+        self,
+        workers=None,
+        scheduler=None,
+        worker=None,
+        asynchronous=False,
+        loop=None,
+        silence_logs=False,
+    ):
+        self._created = weakref.WeakSet()
+        if scheduler is None:
+            try:
+                from distributed.bokeh.scheduler import BokehScheduler
+            except ImportError:
+                services = {}
+            else:
+                services = {("bokeh", 8787): BokehScheduler}
+            scheduler = {"cls": Scheduler, "options": {"services": services}}
+
+        self.scheduler_spec = scheduler
+        self.worker_spec = workers or {}
+        self.new_spec = worker
+        self.workers = {}
+        self._i = 0
+        self._asynchronous = asynchronous
+
+        if silence_logs:
+            self._old_logging_level = silence_logging(level=silence_logs)
+
+        self._loop_runner = LoopRunner(loop=loop, asynchronous=asynchronous)
+        self.loop = self._loop_runner.loop
+
+        self.scheduler = self.scheduler_spec["cls"](
+            loop=self.loop, **self.scheduler_spec["options"]
+        )
+        self.status = "created"
+
+        if not self.asynchronous:
+            self._loop_runner.start()
+            self.sync(self._start)
+
+        self._correct_state_waiting = None
+
+    async def _start(self):
+        while self.status == "starting":
+            await asyncio.sleep(0.01)
+        if self.status == "running":
+            return
+        if self.status == "closed":
+            raise ValueError("Cluster is closed")
+
+        self._lock = asyncio.Lock()
+        self.status = "starting"
+        self.scheduler = await self.scheduler
+        self.status = "running"
+
+    def _correct_state(self):
+        if self._correct_state_waiting:
+            # If people call this frequently, we only want to run it once
+            return self._correct_state_waiting
+        else:
+            task = asyncio.Task(self._correct_state_internal())
+            self._correct_state_waiting = task
+            return task
+
+    async def _correct_state_internal(self):
+        async with self._lock:
+            self._correct_state_waiting = None
+
+            pre = list(set(self.workers))
+            to_close = set(self.workers) - set(self.worker_spec)
+            if to_close:
+                await self.scheduler.retire_workers(workers=list(to_close))
+                tasks = [self.workers[w].close() for w in to_close]
+                for task in tasks:
+                    await task
```

Comment:
Perhaps use `asyncio.gather` or `asyncio.wait` here to avoid a linear iteration when things could be interleaved?

Reply:
Things get a bit tricky because of asyncio/tornado gen.coroutine.  I've fixed this now so that things should work regardless of how `Worker.close` is defined.  Thanks for flagging it.